### ISSUE TITLE
Yarn 2: 404 page does not load

### ISSUE THREAD

**Issue by jonaskuske** [vercel/next.js#11537]
# Bug report

## Describe the bug

When running next using Yarn 2 and visiting a page in the browser that doesn't exist, the 404 error page never shows up. Instead it loads forever ("waiting on localhost...") and then eventually errors with `ERR_EMPTY_RESPONSE`.

But the terminal shows that the page was compiled:

```
[ event ] build page: /next/dist/pages/_error
[ wait ]  compiling ...
[ ready ] compiled successfully - ready on http://localhost:3000
```

## To Reproduce

> I added a repro [here](https://github.com/jonaskuske/next-yarn2-404-repro) – simply clone and run `npm start`, dependencies and yarn itself are included

In the terminal:

```bash
> mkdir repro
> cd repro
> yarn init -i 2
> yarn add next react react-dom
> mkdir pages
> yarn next
```

Then open localhost:3000 in the browser.

## Expected behavior

The browser receives a response and the 404 page is shown.

## System information

- OS: Windows 10 and Ubuntu 16.04 (within Windows Subsystem for Linux 2)
- Browser (if applies): Edge 80, probably all others too
- Version of Next.js: 9.3.3

**Comment by femisco** [comment]
I want to work on this issue can you give me some guidelines please, thanks in advance

**Comment by yellowspaceboots** [comment]
A temporary fix is to create a custom _error.js and 404.js in /pages as described [here](https://nextjs.org/docs/advanced-features/custom-error-page). I was having this same issue where the automatically generated error page from next would not get built correctly using Yarn 2. Once I created my own custom files in /pages it all works with no errors in console. I know it's not a true fix but it should work until a true fix is found. Hope this helps!

**Comment by chabou** [comment]
I worked on it and found some explanations but I am not sure how to really fix it.

onDemandEntries is responsible to create the file on the fly `.next/static/development/pages/next/dist/pages/_error.js`
I don't know exactly why it doesn't use `static/development/pages/_error.js`.
The path is set here:
https://github.com/zeit/next.js/blob/55ffb96a36e9ba61f0e4f556274562f850ab31fb/packages/next/server/on-demand-entry-handler.ts#L283-L286

Due to yarn2, Next fails to generate it. (I don't know exactly why, maybe it depends on `node_modules` ?)
When it fails to find it, the entry is silently deleted from the entries to compile. Because it is assumed that it occurs only if the file has been deleted;
https://github.com/zeit/next.js/blob/55ffb96a36e9ba61f0e4f556274562f850ab31fb/packages/next/server/on-demand-entry-handler.ts#L82-L87

Then the entry is ignored and the handler is not called through the missing call to `doneCallbacks!.emit(page)`:
https://github.com/zeit/next.js/blob/55ffb96a36e9ba61f0e4f556274562f850ab31fb/packages/next/server/on-demand-entry-handler.ts#L163-L181

Finally, the promise in `ensurePage` is not resolved/rejected (because the callback is not called) and the server request is stuck: https://github.com/zeit/next.js/blob/55ffb96a36e9ba61f0e4f556274562f850ab31fb/packages/next/server/on-demand-entry-handler.ts#L308-L337

So I have some questions:
* Do you know why the path of `_error.js` is `.next/static/development/pages/next/dist/pages/_error.js`?
* Should we add a warning and trigger `doneCallbacks!.emit(page)` even when the entry is missing? [here](https://github.com/zeit/next.js/blob/55ffb96a36e9ba61f0e4f556274562f850ab31fb/packages/next/server/on-demand-entry-handler.ts#L168-L170)

Really happy to help.

**Comment by Timer** [comment]
We now have E2E tests for Yarn 2 and it should be working smoothly!

**Comment by joneath** [comment]
I'm currently still facing this issue on Next 10.0.8

**Comment by emrysal** [comment]
Hi @Timer - Wanted to let you know this still seems to be an issue; see this fork I added as a experiment getting Yarn 2 PnP to work with a userspace project. I have added the workaround as suggested by @yellowspaceboots to get around the hanging _error page issue, but in order to reproduce it's just a matter of removing the custom _error.tsx page.

https://github.com/emrysal/calendso/tree/feature/yarn-berry-proof-of-concept

I'm not sure if you're interested in me reporting this; but did so just in case. Thanks to the team for a great project!
